### ISSUE TITLE
calling `write_read` with an empty write buffer repeatedly retries I2C reads

### ISSUE THREAD

**Issue by TheButlah** [esp-rs/esp-hal#201]
With this code on an ESP32-C3:

```rust
i2c
  .write_read(ADDR_DOESNT_MATTER, &[], &mut [0; RECV_LEN_DOESNT_MATTER])
  .expect("Failed i2c");
```

I get this signal trace: [repeated.zip](https://github.com/esp-rs/esp-hal/files/9668043/repeated.zip)

![image](https://user-images.githubusercontent.com/6969415/192863171-ba6f4e22-3421-49c4-86df-15536a678a8a.png)

The code panics due to the `NACK` and then suspends execution in a loop - I have triple checked this with both `probe-rs` and `defmt` logs. So my only conclusion is that the hardware I2C is in a broken state that causes repeated retransmission.

Probably, it is due to this code in `esp-hal`:
![image](https://user-images.githubusercontent.com/6969415/192864083-015b6c92-3ec7-4314-9e58-dd7819c10ced.png)

The write command is issued, but the data to write is never actually set.

**Comment by bjoernQ** [comment]
Thanks for opening this issue!

I did some tests on my side, too.

This is with a no bytes write and something on the bus that ACKs:
![empty_write_ack](https://user-images.githubusercontent.com/5682593/192990074-07d07aea-ef15-4407-b05b-9eeb754b039b.png)

This is with a one bytes write and something that ACKs:
![non_empty_write_ack](https://user-images.githubusercontent.com/5682593/192990312-fd3105dc-1328-4b6c-a01d-4f37ad61253d.png)

This is with a no bytes write and nothing that ACKs: (i.e. I think that is your case)
![empty_write_no_ack](https://user-images.githubusercontent.com/5682593/192990405-3cde6e5d-c81c-4516-b907-b87a68777bd2.png)

This is with a one byte write and nothing that ACKs:
![non_empty_write_no_ack](https://user-images.githubusercontent.com/5682593/192990543-2ae7f7c7-da34-4605-a938-22555f26e9a0.png)

So, what is repeated seems to be the attempt to read.

**Comment by bjoernQ** [comment]
We can get rid of those read attempts by adding `self.reset();` to `execute_transmission` before returning `return Err(Error::AckCheckFailed);` (maybe we should do that in all error cases)

**Comment by bjoernQ** [comment]
Just out of curiosity: is this related to your problem using MPU6050? I ordered a MPU6050 break-out board but which should hopefully arrive by the end of the week

**Comment by TheButlah** [comment]
Yes, it was. I added [this](https://github.com/SlimeVR/SlimeVR-Rust/pull/36/files#diff-0d8effb99b25ba92abf10f742584b05ad99d057c8e6e05055ca8124dc0ee4289R27) PR in which originally I was writing an empty buffer as an attempt to reset any internal state in the I2C peripheral (because IDK what I'm doing lmao)

I now write a single `0x00` instead, and that results in successful communication with the MPU6050

**Comment by jessebraham** [comment]
Can this PR be closed now that #233 has been merged?
